### DISCUSSION TITLE
How to handle errors thrown by generated providers?

### DISCUSSION THREAD

**Discussion by agufagit** [rrousselGit/riverpod#3386] [Q&A] [answered]
for class generated providers, we can use guard() and ref.listen in widget, but what about function generated providers?  how do we handle error and show on snackbar?
```
@riverpod
Stream<List<data>> getData(GetDataRef ref) async* {
    ...
    final remotedata = getFromRemote(...);
     yield remotedata;
    ...
}

final data = ref.watch(getDataProvider);
```

**Comment by Colman** [comment, marked as answer]
Maybe:

```dart
ref.listenSelf((previous, next) {
  if (next.error != null) {
    showSnackbar(...);
  }
});
```

Context might be hard to get inside of your provider. Therefore, you could maybe just add this to your app widget:
```dart
ref.listen(
  getDataProvider,
  (previous, next) {
    if (next.error != null) {
      showSnackbar(...);
    }
  },
);
```

**Comment by agufagit** [comment]
will this work without `guard()` function?

**Reply by Colman** [reply]
I don't think you need to use the guard function in functional providers or the build method of class based providers since they handle setting the async value for you. It only needs to be used in notifier methods that perform side effects. If you don't use guard in these methods, then any errors encountered will be thrown where you called the method from instead of setting the error within the async value.

So to answer your question, yes, it should work within a functional provider without the guard function.
